src/client/Microsoft.Identity.Client/Cache/CacheFallbackOperations.cs

```diff
@@ -298,31 +299,30 @@ internal static class CacheFallbackOperations
         public static MsalRefreshTokenCacheItem GetAdalEntryForMsal(
             ICoreLogger logger,
             ILegacyCachePersistence legacyCachePersistence,
-            string preferredEnvironment,
             IEnumerable<string> environmentAliases,
             string clientId,
             string upn,
             string uniqueId)
         {
-            List<MsalRefreshTokenCacheItem> adalRts = GetAllAdalEntriesForMsal(
+            IEnumerable<MsalRefreshTokenCacheItem> adalRts = GetAllAdalEntriesForMsal(
                 logger,
                 legacyCachePersistence,
                 environmentAliases,
                 clientId,
                 upn,
                 uniqueId);
 
-            List<MsalRefreshTokenCacheItem> filteredByPrefEnv = adalRts.Where
-                (rt => rt.Environment.Equals(preferredEnvironment, StringComparison.OrdinalIgnoreCase)).ToList();
+            IEnumerable<IGrouping<string, MsalRefreshTokenCacheItem>> rtGroupsByEnv = adalRts.GroupBy(rt => rt.Environment.ToLowerInvariant());
 
-            if (filteredByPrefEnv.Any())
+            // if we have more than 1 RT per env, there is smth wrong with the ADAL cache
+            if (rtGroupsByEnv.Any(g => g.Count() > 1))
```

Comment:
ADAL fallback never fails, as we catch all exceptions and log them. But if multiple tokens are found in the cache, it is a legitimate error (someone tampered with the cache) or bug in MSAL

Reply:
I agree. This is why we need this: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/1770